apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/RealReadTransaction.java

```diff
@@ -0,0 +1,36 @@
+package com.apollographql.android.cache.normalized;
+
+import java.util.Collection;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+class RealReadTransaction implements ReadTransaction {
+
+  protected final RealCache cache;
+  protected volatile boolean closed;
+
+  RealReadTransaction(RealCache cache) {
+    this.cache = cache;
+  }
+
+  @Nullable @Override public Record read(@Nonnull String key) {
+    if (closed) {
+      throw new IllegalStateException("Cannot call read on a transaction which has already closed.");
+    }
+    return cache.read(key);
+  }
+
+  @Override public Collection<Record> read(@Nonnull Collection<String> keys) {
```

Comment:
Should it be more generic like:

```
<R> R execute(Transactional<Cache, R> action) {
  return action.call(cache);
}
```

Usage:
```
transaction.execute(new Transactional<>() {
  public Collection<Record> call(Cache cache) {
    return cache.read(keys);  
  }
})
```

Even maybe we can merge Read and Write transactions to one class